### PR TITLE
Initial commits for QEMU FSP mainboard 

### PR DESCRIPTION
**Pull request by rjoleary**
Initial commit for QEMU FSP mainboard

This mainboard will primarily allow us to validate our Rust FFI for FSP
works.

What we have working right now:

* FSP QEMU binary is built and inserted into the image.
* Oreboot currently ignore the FSP binary and boot QEMU without it.

UTK is able to find a valid FSP and it is located at the correct offset:

$ utk target/x86_64-unknown-none/release/image.bin table
Node       GUID/Name                             Type                           Size
BIOS                                                                            0x1000000
 BIOS Pad                                                                       0xf80000
 FV        1B5C27FE-F01C-4FBC-AEAE-341B2E992A17  FFS2                            0x15000
...

Added crate to wrap QEMU FSP blob and header files

* QEMU FSP's source code is in the fspsdk submodule.
* The user is responsible for running
  `git submodule update --init --recursive`.
* The build.rs triggers the Python build system to build the FSP binary
  and header files. bindgen converts the header files to Rust.
* The FSP binary is copied into the crate and compiled into the .fspblob
  ELF section.
* The mainboard's linker script places that section at the fixed offset
  in flash.

### CODE REVIEW

File: src/mainboard/emulation/qemu-fsp/src/romstage.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+use crate::halt;
+use payloads::external::zimage::PAYLOAD;
+
+pub fn romstage() -> ! {
```
Comment (fotonick) [inline src/mainboard/emulation/qemu-fsp/src/romstage.rs:4]: Where is this invoked? I'm probably missing something foundational in how oreboot is put together.

File: src/mainboard/emulation/qemu-fsp/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#[no_mangle]
+pub extern "C" fn _start(fdt_address: usize) -> ! {
+    let io = &mut IOPort;
+    let uart0 = &mut I8250::new(0x3f8, 0, io);
+    uart0.init().unwrap();
+    uart0.pwrite(b"Welcome to oreboot\r\n", 0).unwrap();
+
+    let w = &mut print::WriteTo::new(uart0);
+
+    let kernel_segs = &[payload::Segment { typ: payload::stype::PAYLOAD_SEGMENT_ENTRY, base: 0x80000000, data: &mut SectionReader::new(&Memory {}, 0x20000000 + 0x400000, 6 * 1024 * 1024) }, payload::Segment {
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/src/main.rs:31]: It's due to https://github.com/oreboot/oreboot/blob/master/rustfmt.toml#L5. You'll have to bring it up with Ron; he likes the long lines.

File: src/mainboard/emulation/qemu-fsp/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    write!(w, "Unexpected return from payload\r\n").unwrap();
+    arch::halt()
+}
+
+#[panic_handler]
+fn panic(info: &PanicInfo) -> ! {
+    // Assume that uart0.init() has already been called before the panic.
+    let io = &mut IOPort;
+    let uart0 = &mut I8250::new(0x3f8, 0, io);
+    uart0.pwrite(b"Welcome to oreboot\r\n", 0).unwrap();
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/src/main.rs]: :grinning:

File: src/mainboard/emulation/qemu-fsp/src/romstage.rs (hunk 1 above)
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/src/romstage.rs]: Good question. I also removed https://github.com/oreboot/oreboot/blob/master/src/mainboard/emulation/qemu-q35/src/romstage.rs as well.

It looks like the arm qemu mainboards are using their romstage functions. It'll be worth looking across these qemu targets sometime and refactoring some of the common bits.

File: src/mainboard/emulation/qemu-fsp/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/src/main.rs]: Done. The end goal is not to hard-code these values and instead read them from the fixed-dtfs.dtb file.

File: src/mainboard/emulation/qemu-fsp/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    let w = &mut print::WriteTo::new(uart0);
+    // Printing in the panic handler is best-effort because we really don't want to invoke the panic
+    // handler from inside itself.
+    let _ = write!(w, "PANIC: {}\r\n", info);
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/src/main.rs:47]: This is something I've wondered and don't know how to do idiomatically in Rust. There should be some way to have a global debug print function so you don't have to pass it as an argument to each and every function. Somehow std Rust does this with the println! macro. It looks like there some sort of magic on Stdout to initialize it on first use: https://doc.rust-lang.org/src/std/io/stdio.rs.html#568

[the remark this answers is not in the document]
Comment (fotonick) [reply src/mainboard/emulation/qemu-fsp/src/main.rs:47]: Ah yes, I see your #159. I'll resolve, as I mostly wanted to be sure we were pondering such questions.

**Review by fotonick — DISMISSED** [review]

File: src/mainboard/emulation/qemu-fsp/fixed-dtfs.dts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                compatible = "ore-dtfs";
+                offset = <0xf70000>;
+                size = <0x8000>;
+                file = "$(TARGET_DIR)/fixed-dtfs.dtb";
+            };
+            area@4 {
+                description = "Boot Blob";
+                compatible = "ore-bootblob";
+                offset = <0xf80000>;
+                size = <0x80000>;
```
Comment (rjoleary) [inline src/mainboard/emulation/qemu-fsp/fixed-dtfs.dts:46]: The size of this area is too big for real mode. The rome mainboard has a solution we're going to generalize.

**Review by rminnich — APPROVED** [review]
gorgeous

File: src/mainboard/emulation/qemu-fsp/fixed-dtfs.dts (hunk 5 above)
Comment (rjoleary) [reply src/mainboard/emulation/qemu-fsp/fixed-dtfs.dts:46]: In the meantime, I fixed this by limiting oreboot to 64KiB. It looks like we take up less than half that right now.

**Review by rminnich — DISMISSED** [review]

**Review by rminnich — APPROVED** [review]

**Pull request merged**